packages/openid4vc/src/openid4vc-issuer/repository/SphereonOpenId4VcCredentialOfferSessionStateManager.ts

```diff
@@ -0,0 +1,202 @@
+import type { OpenId4VcIssuanceSessionStateChangedEvent } from '../OpenId4VcIssuerEvents'
+import type { AgentContext } from '@credo-ts/core'
+import type { CredentialOfferSession, IStateManager } from '@sphereon/oid4vci-common'
+
+import { CredoError, EventEmitter } from '@credo-ts/core'
+import { IssueStatus } from '@sphereon/oid4vci-common'
+
+import { OpenId4VcIssuanceSessionState } from '../OpenId4VcIssuanceSessionState'
+import { OpenId4VcIssuerEvents } from '../OpenId4VcIssuerEvents'
+
+import { OpenId4VcIssuanceSessionRecord } from './OpenId4VcIssuanceSessionRecord'
+import { OpenId4VcIssuanceSessionRepository } from './OpenId4VcIssuanceSessionRepository'
+
+export class SphereonOpenId4VcCredentialOfferSessionStateManager implements IStateManager<CredentialOfferSession> {
+  private openId4VcIssuanceSessionRepository: OpenId4VcIssuanceSessionRepository
+  private eventEmitter: EventEmitter
+
+  public constructor(private agentContext: AgentContext, private issuerId: string) {
+    this.openId4VcIssuanceSessionRepository = agentContext.dependencyManager.resolve(OpenId4VcIssuanceSessionRepository)
+    this.eventEmitter = agentContext.dependencyManager.resolve(EventEmitter)
+  }
+
+  public async set(preAuthorizedCode: string, stateValue: CredentialOfferSession): Promise<void> {
+    // Just to make sure that the preAuthorizedCode is the same as the id as that's what we use to query
+    // NOTE: once we support authorized flow, we need to also allow the id to be equal to issuer state
+    if (preAuthorizedCode !== stateValue.preAuthorizedCode) {
+      throw new CredoError('Expected the id of the credential offer state to be equal to the preAuthorizedCode')
+    }
+
+    if (!stateValue.preAuthorizedCode) {
+      throw new CredoError("Expected the stateValue to have a 'preAuthorizedCode' property")
+    }
+
+    // Record may already exist
+    let record = await this.openId4VcIssuanceSessionRepository.findSingleByQuery(this.agentContext, {
+      issuerId: this.issuerId,
+      preAuthorizedCode: stateValue.preAuthorizedCode,
+    })
+
+    const previousState = record?.state ?? null
+
+    let credentialOfferUri = stateValue.credentialOffer.credential_offer_uri
+    if (!credentialOfferUri) {
+      throw new CredoError("Expected the stateValue to have a 'credentialOfferUri' property")
+    }
```

Comment:
Mismatching casing for err message and prop name

Reply:
Ahh yeah. The error message should probably be rewritten. Copilot took the lead here :)

Reply:
Not for now, but I'd say changing the prop name makes more sense. We also change the casing for AnonCreds stuff using class transformer.